### ISSUE TITLE
Is it possible to save/load CL_PROGRAM_BINARIES?

### ISSUE THREAD

**Issue by trxcllnt** [mikeseven/node-opencl#62]
Looking into improving container startup times, I noticed a common pattern is compiling kernels from source once per device, and loading the program binaries from disk on subsequent restarts.

Is it possible to get a handle on the underlying program binaries buffer in order to serialize them to disk, and if so, can the result be loaded via `cl.createProgramWithBinaries()`? Thanks Mike!

**Comment by mikeseven** [comment]
That's one part I've not touched ie containers. I definitely need to update.

Regarding binary, it should work as usual. However, it's not portable. So you would need to generate binaries for each family of GPU. Some would support SPIR and be portable.

What I would do, assuming your kernels don't change, is run them once and get the binary to disk or volume container. Then launch the app again using that binary.

--mike

________________________________

**Comment by trxcllnt** [comment]
Yep, I'm comfortable with that part. We can essentially make it part of the container's `CMD` script.

The part I'm stuck on is getting a reference to the `raw` char* pointer of the `NoCLProgramBinary` instances. The `toString()` value doesn't seem to include the PTX assembly, and there don't seem to be any other methods exposed to v8. Am I missing something?

p.s. I've looked into some other tools like `clcc`, but most I've found aren't maintained and/or segfault on various platforms, so a small node-opencl script seems like the most straightforward if it's possible.

**Comment by mikeseven** [comment]
Yes or could even bind mount a volume.
Indeed, what you hit is the non standard part.
It looks like you are using Nvidia GPU. You should try some of their examples to verify that you can save and load the binary and that it works. I wouldn’t use toString() because that’s really an ArrayBuffer you get back (like all operations with OpenCL). You should get a byte data view, save it and load it, and that will be the same as the native Nvidia example, likely PTX.

**Comment by mikeseven** [comment]
That looks raidonable :-)
Gotta check what the issue is.
Yes in general any data on node-opencl is ArrayBuffer or similar in latest nodejs so data is aligned properly on cpu and gpu memory without incurring translation. I do use other types of buffers for convenience but they are slower because they require translation.

--mike

________________________________
